ls1intum/Artemis
# Feature/programming exercise/Remove automatic build triggers for template repository + trigger the solution on test case update

### Checklist

### Motivation and Context



Fixes steps 1 & 7 in https://github.com/ls1intum/Artemis/issues/772.

This PR addresses a workflow problem that was introduced with the persistance of test cases in the database:
The test cases are generated from the build result of the solution repository. But when e.g. a commit is pushed into the test repository, all repositories are built at the same time. This leads to a problem, because other repository build results might come in before the solution repository; in this case they don't have the updated test cases (e.g. they are missing a new test case).

What we need to achieve here is that we don't create template / student results with outdated test case settings.

### Description


**Implementation:**
1) Remove the build trigger from the test repository to the template repository.
2) Trigger the build of the solution repository manually when a test case is updated (e.g. setting the weight) in `ProgrammingSubmissionService.testCasesChanged`
3) Trigger the template repository, too, when `triggerAll` is executed (before only the studentParticipations).

**What does this mean for legacy exercises?**
- One inconsistency for existing exercises is that now the build for template repositories is executed twice on a push into the test repository: Once because of the Bamboo/Bitbucket config, another time because of the manual trigger we do now in Artemis.
- :warning: As discussed with @krusche we don't create submission anymore on a test repository push. This means that for legacy exercises build results for the student repositories will arrive and have to create submissions with `SubmissionType.OTHER`. This is not a big problem as no programming exercises are running atm, but we will have to keep that in mind. I added extensive docs for this issue.
- We also need to implement removing the build trigger for the new Import feature, as otherwise we would create new programming exercises with the unwanted trigger.

### Steps for Testing


See gif below:
- Push a commit into the test repository, e.g. with a new test
=> The solution participation should now be building; no template / student repository should be building
- Wait for the build result, the solution result should include the new test; there should be no new student or template result

The same should happen when going to Course Administration > Programming Exercise > Manage Test Cases and e.g. updating the weight of a test case and then saving.

### Screenshots



![test-case-trigger](https://user-images.githubusercontent.com/28230611/65907982-b4a0a480-e3c5-11e9-88c5-678991f8cf65.gif)
